### PR TITLE
Add Lazy Loading for cupyx.scipy.* submodules

### PR DESCRIPTION
**Pull request by Raghav323**
PR to add support for lazy loading of cupyx.scipy.* submodules using `__getattr__` as Scipy does .
**Linked Issue**: #7392

### CODE REVIEW

**Comment by Raghav323** [comment]
Could you please review @kmaehashi

File: cupyx/scipy/__init__.py [hunk 1]
```diff
@@ -33,3 +49,12 @@ def get_array_module(*args):
         if isinstance(arg, (_ndarray, _spmatrix)):
             return _cupyx_scipy
     return _scipy
+
+
+def __dir__():
+    return __all__
```
Comment (kmaehashi) [inline cupyx/scipy/__init__.py:55]: This needs to export `get_array_module`.

File: cupyx/scipy/__init__.py [hunk 2]
```diff
@@ -33,3 +49,12 @@ def get_array_module(*args):
...
+
+
+def __getattr__(name):
+    if name in submodules:
+        return _importlib.import_module(f'scipy.{name}')
```
Comment (kmaehashi) [inline cupyx/scipy/__init__.py:60]: This is returning SciPy module instead of CuPy, and not raising an error for undefined attributes.

**Review by kmaehashi — CHANGES_REQUESTED** [review]

**Comment by Raghav323** [comment]
Could you please review @kmaehashi

**Comment by kmaehashi** [comment]
Thanks, but just adding `__getattr__` seems insufficient to support lazy loading, as these `cupyx.scipy` modules are imported automatically via `import cupy`. 

```
$ python -c 'import sys, cupy; print([x for x in sys.modules.keys() if "cupyx.scipy" in x])'
['cupyx.scipy.sparse._util', 'cupyx.scipy.sparse._sputils', 'cupyx.scipy.sparse._base', 'cupyx.scipy.sparse._data', 'cupyx.scipy.sparse._index', 'cupyx.scipy.sparse._compressed', 'cupyx.scipy.sparse._csc', 'cupyx.scipy.sparse._csr', 'cupyx.scipy.sparse._coo', 'cupyxscipy.sparse._dia', 'cupyx.scipy.sparse._construct', 'cupyx.scipy.sparse._extract', 'cupyx.scipy.sparse', 'cupyx.scipy', 'cupyx.scipy.fft._fft', 'cupyx.scipy.special._complexstuff', 'cupyx.scipy.special._trig', 'cupyx.scipy.special._loggamma', 'cupyx.scipy.special._gamma', 'cupyx.scipy.special._bessel', 'cupyx.scipy.special._digamma', 'cupyx.scipy.special._zeta', 'cupyx.scipy.special._gammainc', 'cupyx.scipy.special._beta', 'cupyx.scipy.special._stats_distributions', 'cupyx.scipy.special._statistics', 'cupyx.scipy.special._convex_analysis', 'cupyx.scipy.special._gammaln', 'cupyx.scipy.special._gammasgn', 'cupyx.scipy.special._polygamma', 'cupyx.scipy.special._poch', 'cupyx.scipy.special._erf', 'cupyx.scipy.special._lpmv', 'cupyx.scipy.special._sph_harm', 'cupyx.scipy.special._exp1', 'cupyxscipy.special._expi', 'cupyx.scipy.special._expn', 'cupyx.scipy.special._softmax', 'cupyx.scipy.special._logsoftmax', 'cupyx.scipy.special._basic', 'cupyx.scipy.special._xlogy', 'cupyx.scipy.special._logsumexp', 'cupyx.scipy.special', 'cupyx.scipy.fft._fftlog', 'cupyx.scipy.fft._helper', 'cupyx.scipy.fftpack._fft', 'cupyx.scipy.fftpack', 'cupyx.scipy.fft._realtransforms', 'cupyx.scipy.fft']
```

**Comment by kmaehashi** [comment]
I think this requires some extra work than initially expected in #7392. Let me know if you still want to work on this one.

**Comment by kmaehashi** [comment]
Let me close this one for now as it's been a draft for a while, but feel free to reopen when you have time to resume working on. Thanks!

**Pull request closed without merging**